### PR TITLE
Add new SPECTER2 + SciNCL model

### PR DESCRIPTION
**Pull request by haroldrubio**
This PR adds support for a new model ensemble of SPECTER2 (August 2023 refresh) and SciNCL by pulling models from the HuggingFace model repository.

This PR requires an update to Python 3.8 in order to use the `switch-transformers` library required by SPECTER2.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
           transport.host: localhost
           TZ: "Etc/UTC"
     environment:
-      PYTHON_VERSION: 3.7
+      PYTHON_VERSION: 3.8
```
Comment (melisabok) [inline .circleci/config.yml:28]: which python version is being used in the instances?

File: tests/test_expertise_apiv2.py [hunk 2]
```diff
@@ -394,7 +394,7 @@ def test_submission_content_v2(self, openreview_client, openreview_context, cele
                         'withContent': { 'track': 'no_track' }
                     },
                     "model": {
-                            "name": "specter+mfr",
+                            "name": "specter2+scincl",
```
Comment (melisabok) [inline tests/test_expertise_apiv2.py:397]: is there a chance we can keep a test for both models?

**Comment by melisabok** [comment]
There is test called `test_spectermfr.py` should we create one for this new model?

**Comment by melisabok** [comment]
@carlosmondra please help me to review this PR

File: .circleci/config.yml (hunk 1 above)
Comment (haroldrubio) [reply .circleci/config.yml:28]: The expertise instance is using 3.7 right now so we'd have to do what Carlos did when he updated the matcher's Python version

**Comment by melisabok** [comment]
@haroldrubio can we have running specter+mfr and specter2+scincl running at the same time. I mean, can we ask the user to decide which model to use?

File: expertise/execute_expertise.py [hunk 3]
```diff
@@ -100,6 +100,117 @@ def execute_expertise(config):
...
+            scincl_submissions_path=Path(config['model_params']['submissions_path']).joinpath('sub2vec_scincl.jsonl'),
+            scores_path=Path(config['model_params']['scores_path']).joinpath(config['name'] + '.csv')
+        )
+
+        if config['model_params'].get('sparse_value'):
+            ens_predictor.sparse_scores(
+                scores_path=Path(config['model_params']['scores_path']).joinpath(config['name'] + '_sparse.csv')
+            )
+
+    if config['model'] == 'scincl':
```
Comment (melisabok) [inline expertise/execute_expertise.py:142]: Should we let users to decide to use this model?

I thought we were going to enable 'specter2+scincl' only.

**Comment by melisabok** [comment]
The changes look good to me, I couldn't get in deep details but I have a few questions:

1. The evaluation code is coming from Nihar's paper or something different? Could you add some instructions in the readme file about how to execute the evaluation?
2. The last time we computed scores using this new model, the values were very very high between 0.8 and 1 and we decided to normalize them, is this implemented in this PR?
3. Do we need some instructions about running `specter2+scincl` from command line?

**Comment by carlosmondra** [comment]
> This PR requires an update to Python 3.8 in order to use the `switch-transformers` library required by SPECTER2.

There is still a part of the README that needs to be updated. It reads "This repository only supports Python 3.6 and above.". I'm assuming that should be 3.8 and above or maybe just 3.8.

File: README.md [hunk 4]
```diff
@@ -73,6 +75,32 @@ https://www.overleaf.com/read/ygmygwtjbzfg
 
 https://www.overleaf.com/read/swqrxgqqvmyv
 
+The following instructions are all of the commands to install the dependencies used by this repository - this follows the same commands listed above:
+```
+conda update -y conda
+conda create -n expertise python=$PYTHON_VERSION -c conda-forge
+conda activate expertise
+mkdir ~/expertise-utils
+cd ~/expertise-utils
```
Comment (carlosmondra) [inline README.md:84]: I'm assuming that it's not necessary to create a folder in the home directory or is it?

File: expertise/execute_expertise.py [hunk 5]
```diff
@@ -100,6 +100,117 @@ def execute_expertise(config):
...
+            use_cuda=config['model_params'].get('use_cuda', False),
+            sparse_value=config['model_params'].get('sparse_value'),
+            use_redis=config['model_params'].get('use_redis', False),
+            dump_p2p=config['model_params'].get('dump_p2p', False)
+        )
+        scincl_predictor.set_archives_dataset(archives_dataset)
+        scincl_predictor.set_submissions_dataset(submissions_dataset)
+        scincl_publication_path = Path(config['model_params']['publications_path']).joinpath('pub2vec.jsonl')
+        if config['model_params'].get('use_redis', False):
+            publication_path = None
```
Comment (carlosmondra) [inline expertise/execute_expertise.py:159]: We are not using this anymore right?

File: expertise/models/specter2_scincl/scincl.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                    emb_list.append(paper_emb)
+                except Exception as e:
+                    bad_id_set.add(paper_id)
+
+            emb_tensor = torch.tensor(emb_list, device=torch.device('cpu'))
+            emb_tensor = emb_tensor / (emb_tensor.norm(dim=1, keepdim=True) + 0.000000000001)
+            if bad_id_set:
+                print(f"No Embedding found for {len(bad_id_set)} Papers: ")
+                print(bad_id_set)
+            return emb_tensor, id_list, bad_id_set
```
Comment (carlosmondra) [inline expertise/models/specter2_scincl/scincl.py:233]: I'm guessing this method is useless right now, or am I missing something?

File: expertise/models/specter2_scincl/scincl.py [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+            desc = 'Profiles IDs'
+        for note_id, profile_id, score in tqdm(self.preliminary_scores, total=len(self.preliminary_scores), desc=desc):
+            if counter < self.sparse_value:
+                all_scores.add((note_id, profile_id, score))
+            elif (note_id, profile_id)[id_index] != current_id:
+                counter = 0
+                all_scores.add((note_id, profile_id, score))
+                current_id = (note_id, profile_id)[id_index]
+            counter += 1
+        return all_scores
```
Comment (carlosmondra) [inline expertise/models/specter2_scincl/scincl.py:315]: Is this method being repeated in every model class? Maybe we should start extracting this methods and place them in a common file, unless there is a good reason not to do it.

File: tests/test_expertise_apiv2.py [hunk 8]
```diff
@@ -474,4 +474,103 @@ def test_submission_content_v2(self, openreview_client, openreview_context, cele
...
+            time.sleep(5)
+            response = test_client.get('/expertise/status', query_string={'jobId': f'{job_id}'}).json
+            if response['status'] == 'Error':
+                assert False, response[0]['description']
+            try_time = time.time() - start_time
+
+        assert try_time <= MAX_TIMEOUT, 'Job has not completed in time'
+        assert response['status'] == 'Completed'
+        assert response['name'] == 'test_run'
+        assert response['description'] == 'Job is complete and the computed scores are ready'
```
Comment (carlosmondra) [inline tests/test_expertise_apiv2.py:576]: Should we test that the values are reasonable?

File: tests/test_specter2scincl.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    scores_path = tmp_path / 'scores'
+    scores_path.mkdir()
+    all_scores = specnclModel.all_scores(
+        specter_publications_path=publications_path.joinpath('pub2vec_specter.jsonl'),
+        scincl_publications_path=publications_path.joinpath('pub2vec_scincl.jsonl'),
+        specter_submissions_path=submissions_path.joinpath('sub2vec_specter.jsonl'),
+        scincl_submissions_path=submissions_path.joinpath('sub2vec_scincl.jsonl'),
+        scores_path=scores_path.joinpath(config['name'] + '.csv')
+    )
```
Comment (carlosmondra) [inline tests/test_specter2scincl.py:73]: Why don't check the values of `all_scores` here?

File: expertise/execute_expertise.py (hunk 3 above)
Comment (haroldrubio) [reply expertise/execute_expertise.py:142]: Sorry I just saw this - we don't need to support calling the separate models in openreview-py, this is so I can call them separately from the evaluation code to search over how to combine specter2 and scincl.

File: expertise/models/specter2_scincl/predictor.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from tqdm import tqdm
+class Predictor:
```
Comment (carlosmondra) [inline expertise/models/specter2_scincl/predictor.py:2]: Nice solution!

**Review by carlosmondra — APPROVED** [review]
OK! Thanks for making all the changes. @melisabok I think we can merge this PR.

**Comment by melisabok** [comment]
@haroldrubio is it safe to merge this PR? are we sure we can deploy master and the expertise API will keep working?

**Comment by haroldrubio** [comment]
Yes it should be safe to merge - I'll double check the deployment details with Carlos and plan a deployment during some downtime

**Review by melisabok — APPROVED** [review]
